src/systemctl/systemctl.c

```diff
@@ -515,36 +434,44 @@ static int output_units_list(const UnitInfo *unit_infos, unsigned c) {
                 } else
                         id = u->id;
 
-                if (arg_full) {
-                        e = ellipsize(id, id_len, 33);
-                        if (!e)
-                                return log_oom();
-
-                        id = e;
-                }
-
-                if (circle_len > 0)
-                        printf("%s%s%s ", on_circle, circle ? special_glyph(SPECIAL_GLYPH_BLACK_CIRCLE) : " ", off_circle);
+                r = table_add_many(table,
+                                   TABLE_STRING, circle ? special_glyph(SPECIAL_GLYPH_BLACK_CIRCLE) : " ",
+                                   TABLE_SET_COLOR, on_circle,
+                                   TABLE_STRING, id,
+                                   TABLE_SET_COLOR, on_active,
+                                   TABLE_STRING, u->load_state,
+                                   TABLE_SET_COLOR, on_loaded,
+                                   TABLE_STRING, u->active_state,
+                                   TABLE_SET_COLOR, on_active,
+                                   TABLE_STRING, u->sub_state,
+                                   TABLE_SET_COLOR, on_active,
+                                   TABLE_STRING, u->job_id ? u->job_type: "",
+                                   TABLE_SET_COLOR, u->job_id ? on_underline : "",
+                                   TABLE_STRING, u->description,
+                                   TABLE_SET_COLOR, on_underline);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to insert a row: %m");
 
-                printf("%s%s%-*s%s %s%-*s%s %s%-*s %-*s%s %-*s",
-                       on_underline,
-                       on_active, id_len, id, off_active,
-                       on_loaded, load_len, u->load_state, off_loaded,
-                       on_active, active_len, u->active_state,
-                       sub_len, u->sub_state, off_active,
-                       job_count ? job_len + 1 : 0, u->job_id ? u->job_type : "");
+                if (u->job_id != 0)
+                        job_count++;
+        }
 
-                printf("%-*.*s%s\n",
-                       desc_len,
-                       full ? -1 : (int) desc_len,
-                       u->description,
-                       off_underline);
+        if (job_count == 0) {
+                /* There's no data in the JOB column, so let's hide it */
+                r = table_set_display(table, 0, 1, 2, 3, 4, 6, (size_t) -1);
```

Comment:
This part right now is actually a problem and it took me a while to find out why.

The CentOS CI started segfaulting due to this error:

```
[  347.857026] sh[72]: + systemctl --state=failed --no-legend --no-pager
[  347.901645] sh[73]: Assertion 'column < t->n_columns' failed at src/shared/format-table.c:979, function int table_set_display(Table *, size_t, ...)(). Aborting.
```

After a while I was able to reproduce it by dropping the condition around `r = table_set_display(table, 0, 1, 2, 3, 4, 6, (size_t) -1);`, but I still couldn't comprehend why it fails. After few `fprintf()` statements later, the issue seems pretty obvious:

```patch
diff --git a/src/shared/format-table.c b/src/shared/format-table.c
index 4617ae8bad..e6f7987e46 100644
--- a/src/shared/format-table.c
+++ b/src/shared/format-table.c
@@ -975,7 +975,9 @@ int table_set_display(Table *t, size_t first_column, ...) {
         column = first_column;
 
         va_start(ap, first_column);
+        fprintf(stderr, "n_columns: %lu\n", t->n_columns);
         for (;;) {
+                fprintf(stderr, "column: %lu\n", column);
                 assert(column < t->n_columns);
 
                 if (!GREEDY_REALLOC(t->display_map, allocated, MAX(t->n_columns, t->n_display_map+1))) {
@@ -985,10 +987,14 @@ int table_set_display(Table *t, size_t first_column, ...) {
 
                 t->display_map[t->n_display_map++] = column;
 
+                fprintf(stderr, "column before change: %lu\n", column);
                 column = va_arg(ap, size_t);
+                fprintf(stderr, "column after change: %lu\n", column);
                 if (column == (size_t) -1)
                         break;
 
+                fprintf(stderr, "column after cond: %lu\n", column);
+
         }
         va_end(ap);
```

Output:
```
$ build/systemctl --user --state=failed 
n_columns: 7
column: 0
column before change: 0
column after change: 1
column after cond: 1
column: 1
column before change: 1
column after change: 2
column after cond: 2
column: 2
column before change: 2
column after change: 3
column after cond: 3
column: 3
column before change: 3
column after change: 4
column after cond: 4
column: 4
column before change: 4
column after change: 140733193388038
column after cond: 140733193388038
column: 140733193388038
Assertion 'column < t->n_columns' failed at src/shared/format-table.c:981, function int table_set_display(Table *, size_t, ...)(). Aborting.
Aborted (core dumped)
```

After casting all numbers to `size_t`, it works as expected. I tried making a custom enum for these values, as is already done [here](https://github.com/systemd/systemd/blob/master/src/busctl/busctl.c#L210) but that didn't work either.

So, is there any nicer solution than typecasting all `table_set_display()` values to `size_t`?

Reply:
yikes. I am not aware of any. Bummer.

Reply:
Fair enough, let's stick with `size_t` typecast then.